developing-without-a-build/article.md

```diff
@@ -0,0 +1,257 @@
+<script type="module">
+  import './app.js';
+
+  console.log('hello world!');
+</script>
+```
+
+Note that we need to use an explicit file extension, as the browser doesn't know which file to request otherwise.
+
+The same thing works if we use the `src` attribute:
+
+```html
+<script type="module" src="./app.js"></script>
+```
+
+### Loading dependencies
+We don't write our code in just one file. After importing the initial module, we can import other modules. For example let's create two new files:
+
+`src/app.js`:
+```javascript
+import { message } from './message.js';
+
+console.log(`The message is: ${message}`);
+```
+
+`src/message.js`:
+```js
+export const message = 'hello world';
+```
+
+Place both files in a `src` directory and import `app.js` from your index.html:
+
+```html
+<!DOCTYPE html>
+<html>
+<head></head>
+
+<body>
+  <script type="module" src="./src/app.js"></script>
+</body>
+
+</html>
+```
+
+If you run this and check the network panel, you will see both modules being loaded. Because imports are resolved relatively, `app.js` can refer to `message.js` using a relative path:
+
+![](./images/modules.png)
+
+This seems trivial, but it is extremely useful and something we did not have before with classic scripts. We no longer need to coordinate dependencies somewhere central or maintain a base url. Modules can declare their own dependencies, and we can import any module without knowing what their dependencies are. The browser takes care of requesting the correct files.
+
+### Dynamic imports
+When building any serious web application we are usually going to need to do some form of lazy loading for best performance. With the browser's module system we can do this using dynamic imports. Static imports like we saw before always need to exist on the top level.
+
+For example we cannot write:
+```javascript
+if (someCondition) {
+  import './bar.js';
+}
+```
+
+This is what dynamic imports are for. Dynamic imports can import a module at any time. It returns a promise that resolves with the imported module.
+
+For example let's update the `app.js` example we created above:
+```javascript
+window.addEventListener('click', async () => {
+  const module = await import('./message.js');
+
+  console.log(`The message is: ${module.message}`);
+});
+```
+
+Now we are not importing the message module right away, but delaying it until the user has clicked anywhere on the page. We can await the promise returned from the import and interact with the modules. Any exported members are avaible on the module object.
+
+### Lazy evaluation
+This is where developing without a bundler has a significant benefit. If you bundle your application before serving it to the browser, the bundler needs to evaluate all your dynamic imports to do code splitting and output separate chunks. For large applications with a lot of dynamic imports this can add significant overhead as the entire application is built and bundled before you can see anything anything.
```

Comment:
```suggestion
This is where developing without a bundler has a significant benefit. If you bundle your application before serving it to the browser, the bundler needs to evaluate all your dynamic imports to do code splitting and output separate chunks. For large applications with a lot of dynamic imports this can add significant overhead as the entire application is built and bundled before you can see anything.
```